### PR TITLE
Check for use of modifiers in invalid contexts

### LINKED ISSUE #7550: ICE on missing modifier identifier
## Description

Initially reported by @CHM1993.

The contract below:

```
contract test {
    modifier mod() { _; }

    function f() public {
        mod  ;
    }
}
```
reports an internal compiler error:
```
Internal compiler error during compilation:
/build/solc/parts/solidity/src/libsolidity/codegen/ExpressionCompiler.cpp(1672): Throw in function virtual void dev::solidity::ExpressionCompiler::endVisit(const dev::solidity::Identifier&)
Dynamic exception type: boost::exception_detail::clone_impl<langutil::InternalCompilerError>
std::exception::what: Identifier type not expected in expression context.
[dev::tag_comment*] = Identifier type not expected in expression context.
```

## Environment

- Compiler version: 0.5.11+
- Target EVM version (as per compiler settings): 0.5.11+commit.22be8592.Linux.g++
- Framework/IDE (e.g. Truffle or Remix): Remix (i found it from command line, while fuzzing the solc compiler)
- EVM execution environment / backend / blockchain client:
- Operating system: Ubuntu 18.04

## Steps to Reproduce

Run `./build/test/tools/solfuzzer`

### PR DESCRIPTION
**Pull request by Marenz**
fixes #7550

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -2376,15 +2381,24 @@ bool TypeChecker::visit(Identifier const& _identifier)
...
+				"\"suicide\" has been deprecated in favour of \"selfdestruct\"."
 			);
 	}
 
+	if (!m_insideModifierInvocation)
+		if (ModifierType const* type = dynamic_cast<decltype(type)>(_identifier.annotation().type))
+		{
+			m_errorReporter.typeError(
+				_identifier.location(),
+				"Modifier can only be referenced in function signatures."
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: A function signature is a string representing the type of the function, and not a syntactical construct. This should maybe read "function headers".

**Comment by chriseth** [comment]
I think it would be better to do that inside the syntax checker instead of the type checker - it is responsible for validating that certain elements of the AST are in certain places.

**Comment by chriseth** [comment]
Please add a changelog entry.

**Comment by Marenz** [comment]
> I think it would be better to do that inside the syntax checker instead of the type checker - it is responsible for validating that certain elements of the AST are in certain places.

Repeating what I said in the chat: This is problematic as we don't have the type information at that stage

**Comment by chriseth** [comment]
Conceptually, this is not about types but about kinds of declarations. So we do not need types but just resolved identifiers. I think we should take care not to clutter the state of TypeChecker - at best it should not have any state.

Another fix would be to just ignore the type in the code generator - it has a proper type and we might generate a "statement does not have an effect" error there.

**Comment by Marenz** [comment]
The declaration itself is also not set at this stage.

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
Will merge this but also created a new issue to clean it up: https://github.com/ethereum/solidity/issues/7566

**Comment by chriseth** [comment]
Ah, this is missing a changelog entry.

**Pull request merged**